src/jupyterhub_config/custom_docker_spawner.py

```diff
@@ -41,6 +44,59 @@ def start(self):
 
         return super().start()
 
+    def _get_idle_timeout(self):
+        """
+        Retrieves the idle timeout from the environment variable `IDLE_TIMEOUT_MINUTES`.
+        If not set, defaults to 180 minutes.
+
+        Returns:
+            timedelta: Idle timeout duration.
+        """
+        idle_timeout_minutes = int(os.getenv("IDLE_TIMEOUT_MINUTES", self.DEFAULT_IDLE_TIMEOUT_MINUTES))
+        self.log.info(f"Idle timeout set to {idle_timeout_minutes} minutes")
+        return timedelta(minutes=idle_timeout_minutes)
+
+    async def poll(self):
+        """
+        Overrides the poll method to periodically checks the status of the user’s JupyterHub container.
+
+        ref:
+        https://github.com/jupyterhub/dockerspawner/blob/main/dockerspawner/dockerspawner.py#L1004
+        https://jupyterhub-dockerspawner.readthedocs.io/en/latest/api/index.html#dockerspawner.DockerSpawner.poll
+
+        - If the container is stopped, returns the status immediately.
+        - If the container is running, checks how long the user has been idle.
+        - If idle time exceeds the defined threshold, stops the container to save resources.
+
+        The poll method is invoked at regular intervals by the Spawner, with the frequency determined by the JupyterHub
+        server's configuration (default is 30 seconds).
+
+        Returns:
+            int or None: Returns an exit code (0) if the container has been stopped due
+                         to inactivity. Returns None if the container is still active
+                         and running.
+        """
+        # Check if the container has already stopped
+        status = await super().poll()
+        if status is not None:
+            # Container has already stopped, return its status code immediately
+            return status
+
+        last_activity = self.user.last_activity
```

Comment:
So this must be running in the central hub server since it's polling against possibly stopped containers. Is there a CustomDockerSpawner instance per user?

Reply:
:+1: